src/bin/cql-stress-cassandra-stress/operation/mod.rs

```diff
@@ -310,3 +314,63 @@ impl RowValidator for ExistsRowValidator {
         Ok(())
     }
 }
+
+/// A sampler created based on a ratio map and a counter distribution.
+///
+/// How the sampler works?
+/// One iteration consists of:
+/// - sampling an item based on ratio map. `current_item_index` is sampled from `item_index_dist`.
+///   The item can then be retrieved via this index from `items` vector.
+/// - sampling a counter which says how many times to return the current item.
+///   The counter is sampled from `counter_dist` distribution.
+///
+/// The user then can sample the items via `sample` or `sample_cached` method.
+///
+/// The `sample` method will decrease the counter by 1, and return current item.
+/// If the counter reaches 0, new iteration starts.
+///
+/// The `sample_cached` method returns a current item without decreasing the counter.
+/// This is helpful when the user wants to, for example, retry an operation that was
+/// sampled before, but failed for some reason.
+#[cfg(feature = "user-profile")]
+struct OperationSampler<T> {
+    counter_dist: Box<dyn Distribution>,
+    items: Vec<T>,
+    item_index_dist: WeightedIndex<f64>,
+    current_item_remaining: u8,
+    current_item_index: usize,
+}
+
+#[cfg(feature = "user-profile")]
+impl<T> OperationSampler<T> {
+    pub fn new(
+        weights: impl Iterator<Item = (T, f64)>,
+        counter_dist_factory: &dyn DistributionFactory,
+    ) -> Self {
+        let (items, weights): (Vec<_>, Vec<_>) = weights.unzip();
+        // We verify the ratio properties during parsing.
+        let item_index_dist = WeightedIndex::new(weights)
+            .expect("Failed to create a WeightedIntex from provided ratios");
+
+        Self {
+            counter_dist: counter_dist_factory.create(),
+            items,
+            item_index_dist,
+            current_item_remaining: 0,
+            current_item_index: 0,
+        }
+    }
+
+    pub fn sample(&mut self) -> &T {
+        if self.current_item_remaining == 0 {
+            self.current_item_index = self.item_index_dist.sample(&mut rand::thread_rng());
+            self.current_item_remaining = (self.counter_dist.next_i64() as u8).max(1);
+        }
+        self.current_item_remaining -= 1;
+        &self.items[self.current_item_index]
+    }
+
+    pub fn sample_cached(&self) -> &T {
+        &self.items[self.current_item_index]
+    }
```

Comment:
I would rather name this function `previous_sample`, because actually no sampling happens here - just returning what was returned previously.
[_Probability sampling refers to the selection of a sample from a population, when this selection is based on the principle of randomization, that is, random selection or chance._](https://www150.statcan.gc.ca/n1/edu/power-pouvoir/ch13/prob/5214899-eng.htm)
...and there is no random choice here.